openapi3filter/middleware.go

```diff
@@ -0,0 +1,272 @@
+			return
+		}
+
+		var wr responseWrapper
+		if v.strict {
+			wr = &strictResponseWrapper{w: w}
+		} else {
+			wr = &warnResponseWrapper{w: w}
+		}
+
+		h.ServeHTTP(wr, r)
+
+		err = ValidateResponse(r.Context(), &ResponseValidationInput{
+			RequestValidationInput: requestValidationInput,
+			Status:                 wr.statusCode(),
+			Header:                 wr.Header(),
+			Body:                   nopCloser{bytes.NewBuffer(wr.bodyContents())},
+		})
+		if err != nil {
+			v.logFunc("invalid response", err)
+			if v.strict {
+				v.errFunc(w, http.StatusInternalServerError, ErrCodeResponseInvalid, err)
+			}
+			return
+		}
+
+		err = wr.flushBodyContents()
+		if err != nil {
+			v.logFunc("failed to write response", err)
+		}
+	})
+}
+
+type nopCloser struct {
+	io.Reader
+}
+
+// Close implements io.Closer.
+func (nopCloser) Close() error {
+	return nil
+}
+
+type responseWrapper interface {
+	http.ResponseWriter
+
+	// flushBodyContents writes the buffered response to the client, if it has
+	// not yet been written.
+	flushBodyContents() error
+
+	// statusCode returns the response status code, 0 if not set yet.
+	statusCode() int
+
+	// bodyContents returns the buffered
+	bodyContents() []byte
+}
+
+type warnResponseWrapper struct {
+	w      http.ResponseWriter
+	fl     http.Flusher
+	status int
+	body   bytes.Buffer
+}
+
+// Write implements http.ResponseWriter.
+func (l *warnResponseWrapper) Write(b []byte) (int, error) {
+	if l.status == 0 {
+		l.w.WriteHeader(l.status)
+	}
+	n, err := l.w.Write(b)
+	if err == nil {
+		l.body.Write(b)
+	}
+	return n, err
+}
+
+// WriteHeader implements http.ResponseWriter.
+func (l *warnResponseWrapper) WriteHeader(status int) {
+	if l.status == 0 {
+		l.status = status
+	}
+	l.w.WriteHeader(l.status)
+}
+
+// Header implements http.ResponseWriter.
+func (wr *warnResponseWrapper) Header() http.Header {
+	return wr.w.Header()
+}
+
+// Flush implements the optional http.Flusher interface.
+func (wr *warnResponseWrapper) Flush() {
+	if fl, ok := wr.w.(http.Flusher); ok {
+		fl.Flush()
+	}
+}
+
+func (l *warnResponseWrapper) flushBodyContents() error {
+	return nil
+}
+
+func (l *warnResponseWrapper) statusCode() int {
+	return l.status
+}
+
+func (l *warnResponseWrapper) bodyContents() []byte {
+	return l.body.Bytes()
+}
+
+type strictResponseWrapper struct {
+	w      http.ResponseWriter
+	status int
+	body   bytes.Buffer
+}
+
+// Write implements http.ResponseWriter.
+func (wr *strictResponseWrapper) Write(b []byte) (int, error) {
+	if wr.status == 0 {
+		wr.status = http.StatusOK
+	}
+	return wr.body.Write(b)
+}
+
+// WriteHeader implements http.ResponseWriter.
+func (wr *strictResponseWrapper) WriteHeader(status int) {
+	if wr.status == 0 {
+		wr.status = status
+	}
+}
+
+// Header implements http.ResponseWriter.
+func (wr *strictResponseWrapper) Header() http.Header {
+	return wr.w.Header()
+}
+
+func (wr *strictResponseWrapper) flushBodyContents() error {
+	wr.w.WriteHeader(wr.status)
```

Comment:
As per above 
```suggestion
```

Reply:
Not sure if this was clear (and if godoc comments could explain it better, please let me know!): we don't want to flush the response through to the underlying ResponseWriter in strict mode, because the response may be out of compliance with OpenAPI and invalid; we'll want to `wr.w.WriteHeader(500)` (or give our errFunc the chance to) when that happens in strict mode. If we pass through the call to WriteHeader, we won't be able to override it.